doc/content/demos/multi-user-chat/multi-user-chat.py

```diff
@@ -0,0 +1,205 @@
+            ),
+
+            TextInput(
+                placeholder='Name',
+                _id='new-room-name',
+                bubble_up=True,
+            ),
+            Button('Create', _id='create-new-room'),
+        )
+
+        # choose a room
+        while True:
+            input_event = self.await_input_event(html=html)
+
+            # button
+            if input_event.node_has_id('new-room-name'):
+                new_room = input_event.node.value
+
+                if new_room in rooms:
+                    html.query_selector('#create-new-room').disabled = True
+
+                else:
+                    html.query_selector('#create-new-room').disabled = False
+
+            # textinput
+            if input_event.node_has_id('create-new-room'):
+                new_room = html.query_selector('#new-room-name').value
+
+                return {
+                    'redirect': f'/{new_room}',
+                }
+
+
+@app.route('/<room>(/)?')
+class Chat(LonaView):
+
+    # messages ################################################################
+    def show_message(self, message_data):
+        message = Span(message_data['message'])
+
+        if message_data['type'] == 'action':
+            message.style['font-weight'] = 'bold'
+
+        self.messages.append(
+            Div(Span(f'{message_data["issuer"]}: '), message),
+        )
+
+    def on_view_event(self, view_event):
+        # view is not ready yet
+        if not hasattr(self, 'fully_set_up'):
+            return
+
+        # message is for another room
+        if view_event.data['room'] != self.room_name:
+            return
+
+        self.show_message(view_event.data)
+
+    def send_message(self, message, message_type):
+        message_data = {
+            'room': self.room_name,
+            'issuer': self.name,
+            'message': message,
+            'type': message_type,
+        }
+
+        self.fire_view_event('message', message_data)
+
+        self.server.state['rooms'][self.room_name].append(message_data)
+
+    # gui #####################################################################
+    def handle_input_event(self, input_event):
+        if not input_event.node_has_id('send'):
+            return input_event
+
+        if not self.text_area.value:
+            return
+
+        self.send_message(self.text_area.value, message_type='chat')
+        self.text_area.value = ''
+
+    def choose_name(self):
+        name = TextInput(placeholder='Name')
+        message = P(_style='color: red')
+
+        html = HTML(
+            H1(self.room_name),
+            P('Pick a name'),
+            message,
+            name,
+            Button('Join'),
+        )
+
+        while True:
+            self.await_click(html=html)
+
+            if not name.value:
+                continue
+
+            # check if name is already taken for this room
+            name_already_taken = False
+
+            for view in self.iter_objects():
```

Comment:
nit: I think it's better to store user names in server state instead of iterating over the views.

Reply:
I absolutely agree, but usernames then would be taken indefinitely because  currently there is no good way to react to a user disconnecting from a view (i currently prepare a PR to fix that)